src/compression/b44/mod.rs

```diff
@@ -0,0 +1,924 @@
+    assert_eq!(b.len(), 14);
+    assert_ne!(b[2], 0xfc);
+
+    s[0] = ((b_u32(&b, 0) << 8) | b_u32(&b, 1)) as u16;
+
+    let shift = (b_u32(&b, 2) >> 2);
+    let bias = 0x20 << shift;
+
+    s[4] = (s[0] as u32 + ((((b_u32(&b, 2) << 4) | (b_u32(&b, 3) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[8] = (s[4] as u32 + ((((b_u32(&b, 3) << 2) | (b_u32(&b, 4) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[12] = (s[8] as u32 + ((b_u32(&b, 4) & SIX_BITS) << shift) - bias) as u16;
+
+    s[1] = (s[0] as u32 + ((b_u32(&b, 5) >> 2) << shift) - bias) as u16;
+    s[5] = (s[4] as u32 + ((((b_u32(&b, 5) << 4) | (b_u32(&b, 6) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[9] = (s[8] as u32 + ((((b_u32(&b, 6) << 2) | (b_u32(&b, 7) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[13] = (s[12] as u32 + ((b_u32(&b, 7) & SIX_BITS) << shift) - bias) as u16;
+
+    s[2] = (s[1] as u32 + ((b_u32(&b, 8) >> 2) << shift) - bias) as u16;
+    s[6] = (s[5] as u32 + ((((b_u32(&b, 8) << 4) | (b_u32(&b, 9) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[10] = (s[9] as u32 + ((((b_u32(&b, 9) << 2) | (b_u32(&b, 10) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[14] = (s[13] as u32 + ((b_u32(&b, 10) & SIX_BITS) << shift) - bias) as u16;
+
+    s[3] = (s[2] as u32 + ((b_u32(&b, 11) >> 2) << shift) - bias) as u16;
+    s[7] = (s[6] as u32 + ((((b_u32(&b, 11) << 4) | (b_u32(&b, 12) >> 4)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[11] = (s[10] as u32 + ((((b_u32(&b, 12) << 2) | (b_u32(&b, 13) >> 6)) & SIX_BITS) << shift)
+        - bias) as u16;
+    s[15] = (s[14] as u32 + ((b_u32(&b, 13) & SIX_BITS) << shift) - bias) as u16;
+
+    for i in 0..16 {
+        if (s[i] & 0x8000) != 0 {
+            s[i] &= 0x7fff;
+        } else {
+            s[i] = !s[i];
+        }
+    }
+}
+
+// Unpack a 3-byte block into 4 by 4 identical 16-bit pixels.
+fn unpack3(b: &[u8], s: &mut [u16; 16]) {
+    // TODO: b slice should be &mut [u8; 3], but rust doesn't support sized slice.
+    assert_eq!(b[2], 0xfc);
+
+    s[0] = (((b[0] as u32) << 8) | (b[1] as u32)) as u16;
+
+    if (s[0] & 0x8000) != 0 {
+        s[0] &= 0x7fff;
+    } else {
+        s[0] = !s[0];
+    }
+
+    for i in 0..16 {
+        s[i] = s[0];
+    }
+}
+
+#[derive(Debug)]
+struct ChannelData {
+    tmp_start_index: usize,
+    tmp_end_index: usize,
+
+    resolution: Vec2<usize>,
+    y_sampling: usize,
+    type_: SampleType,
+    quantize_linearly: bool,
+    samples_per_pixel: usize,
+}
+
+fn cpy_u16(src: &[u16], src_i: usize, dst: &mut [u16], dst_i: usize, n: usize) {
+    // assert_eq!(src.len(), dst.len());
+
+    // for i in 0..src.len() {
+    for i in 0..n {
+        dst[dst_i + i] = src[src_i + i];
+    }
+}
+
+pub fn decompress(
+    channels: &ChannelList,
+    compressed: &ByteVec,
+    rectangle: IntegerBounds,
+    expected_byte_size: usize,
+    _pedantic: bool,
```

Comment:
What `_pedantic` is supposed to do?

Reply:
This is a flag the user sets when reading or writing the image. If pedantic is true, then an error will be returned as soon as anything is missing in the file, or two values in the image contradict each other. If pedantic is false, then only fatal errors will be thrown. By default, reading an image is not pedantic, which means that slightly invalid files might still be readable. For example, if some attribute is missing but can be recomputed, this flag decides whether an error is thrown. Or if the pedantic flag is true and there are still bytes left after the decompression algorithm finished, an error is thrown, because this should not happen and something might be wrong with the file. Or if your application is a target of attacks, or if you want to emulate the original C++ library, you might want to switch to pedantic reading. 

When writing an image, the pedantic flag can be used to skip the validation of your image. For example, if you write an uncompressed file and need maximum speed, it might save a millisecond to disable the checks, if you know that your file is not invalid any ways. I do not recommend this though, as the file might not be readably by any other exr library after that.

In the other compression methods, I think it's only used in the decompression methods, where the buffer is checked after decompression should be over. When bytes are still left even though the decompression is over, and pedantic is true, then an error is returned. If you think this is a simple addition, then you could also do this for B44.